src/librustc_interface/passes.rs

```diff
@@ -1015,7 +1015,7 @@ pub fn start_codegen<'tcx>(
     tcx: TyCtxt<'tcx>,
     outputs: &OutputFilenames,
 ) -> Box<dyn Any> {
-    if log_enabled!(::log::Level::Info) {
+    if !::log::info_span!("").is_none() {
         println!("Pre-codegen");
         tcx.print_debug_stats();
     }
```

Comment:
tracing doesn't have something equivalent to `log_enabled!`, as control over where a log/span is enabled is the responsibility of the Subscriber, not the lexically-scoped callsite.

Additionally, `info_span!("").is_none()` will _always_ evaluate to false, meaning this conditional will _always_ be true. During this transition period, consider using [`Span::is_disabled`](https://docs.rs/tracing/0.1.17/tracing/struct.Span.html#method.is_disabled) instead:

```rust
// assume that `RUST_LOG` is equivalent to:
tracing_subscriber::fmt()
    // all spans/events with a level higher than DEBUG (e.g, info, warn, etc.)
    // will be written to stdout.
    .with_max_level(Level::DEBUG)
    // sets this to be the default, global subscriber for this application.
    .init();

// elsewhere, like here...
if !::tracing::trace_span!("").is_disabled() {
    // with the above settings, this shouldn't ever be evaluated!
    println!("Pre-codegen");
    tcx.print_debug_stats();
}
```

Reply:
I _suspect_ that this is the source of the poor perf report.

Reply:
Yeah, `is_none` and `is_disabled` have different semantics.

 [`is_disabled`](https://docs.rs/tracing/0.1.17/tracing/span/struct.Span.html#method.is_disabled):
> Returns true if this span was disabled by the subscriber and does not exist.

[`is_none`](https://docs.rs/tracing/0.1.17/tracing/span/struct.Span.html#method.is_none):
>  Returns `true` if this span was constructed by `Span::none` and is empty.
>
> If `is_none` returns `true` for a given span, then `is_disabled` will
> also return `true`. However, when a span is disabled by the subscriber
> rather than constructed by `Span::none`, this method will return
> `false`, while `is_disabled` will return `true`.

So, `is_none` _only_ returns `true` if the span was constructed by an explicit call to the `Span::none()` function. Since these spans were created by a macro and may be disabled by the subscriber, they will never return `true` for calls to `is_none`. That means that all the code that previously was only run at the trace level is no always being run.

I think there may also be _some_ minor additional performance cost to creating these empty spans in the case where they are _not_ disabled, but I don't think that's as big an issue. The best solution for _that_ would probably be eventually refactoring this code to not rely on empty spans to test if a given level/module combination is enabled.

Reply:
I was reading the docs and bodies of both, and this should have made sense to me, but I guess I completely short circuited to thinking they were each doing the what the other thing was doing.

Reply:
> The best solution for that would probably be eventually refactoring this code to not rely on empty spans to test if a given level/module combination is enabled.

Yea, I want to create debug printers for all these situations which will then only be invoked in an `event!` invocation instead of running code conditionally depending on the current tracing level.

Reply:
> > The best solution for that would probably be eventually refactoring this code to not rely on empty spans to test if a given level/module combination is enabled.
> 
> Yea, I want to create debug printers for all these situations which will then only be invoked in an `event!` invocation instead of running code conditionally depending on the current tracing level.

Yup, I was thinking about opening a PR against this branch to do that...but if you're already planning on it, that would save me from having to figure out how to compile rustc. :)

Reply:
I've done this in https://github.com/rust-lang/rust/pull/74876